### ISSUE TITLE
Incorrect Docker Version Detected when RPMs Preinstalled

### ISSUE THREAD

**Issue by cpitman** [openshift/openshift-ansible#1958]
We have some clients that want to do disconnected/offline installs without also setting up an RPM repository. In that case we're downloading the rpms and then installing them on the hosts to be deployed manually then running the installer.

The installer then tries to determine the docker version by looking at what version of openshift is in subscribed repos. Since the machine is subscribed to nothing, this defaults to docker 1.8 even though we're installed 3.2.

This seams like a corner case, but at a minimum it would be useful if we could override the detected version from the inventory.

**Comment by sdodson** [comment]
Rather than having the installer short circuit all the code paths that may rely on a yum repo why not create a yum repo in the disconnected environment and then add that repo to all hosts via openshift_additional_repos ?

**Comment by stratus-ss** [comment]
Some businesses may not allow a repository to be present in the disconnected environment. There is a specific use case, which like @cpitman said is probably an edge case.

In the end it could be worked around it by editing the playbooks/byo/roles/openshift_docker_facts/tasks/main.yaml file but this takes more knowledge to implement then simply an option in the ansible installer

**Comment by sdodson** [comment]
You're already copying the rpms to the host, all you'd have to do is run `createrepo` in that directory and have the installer use that directory.

**Comment by stratus-ss** [comment]
I'm not totally disagreeing with your point of view. However I would point out that in the scenario that client does not allow apache to host a repo, you would then have to

1) push the rpms
2) create a repo
3) create the repo file
4) update the box

If there is a large environment this is more tedious than simply installing the rpms. It is definitely something to consider writing a playbook for.

Simplest solution from my uneducated point of view is to allow an override for disconnected installs, while you cannot account for all cases when programming, the repoquery method makes some very specific assumptions which may not be true in all cases.

Is the simplest solution the best one? Honestly, I don't think I am qualified to even have an opinion on that (in public). I can only express scenarios I have to work around. If it is deemed that this is too edge case, there is a workaround that should be documented for such edge cases

**Comment by detiber** [comment]
@etsauer What are your thoughts here? How does this fit into the CASL view of disconnected installs?

**Comment by detiber** [comment]
@cpitman @stratus-ss: We've been working with @etsaur and the CASL team on gathering requirements for disconnected installations (and collaboratively working on implementing those requirements), It might be a good idea to sync up with them to discuss how your use cases might differ from the ones they have been working towards.

**Comment by etsauer** [comment]
@detiber the CASL approach and general consulting approaches we discussed last month assumed you are allowed host a remote yum repo somewhere in your internal environment. @cpitman's case is definitely more unique.

As to @stratus-ss's point of having 4 steps being tedious for large environments, I would argue that step 1 (push rpms) would have to be done for all hosts no matter what and is the most tedious which would mean that it would need to be automated anyway. I don't see that as being a big deal.

On the other hand, if the change was a simple one to make and it helped multiple customers it's probably worth doing.

**Comment by detiber** [comment]
@dgoodwin any thoughts on how difficult it would be to opt out of the package detection via flag with your upcoming changes?

**Comment by sdodson** [comment]
Yeah, it should certainly get easier in the near future.

**Comment by dgoodwin** [comment]
Yeah I believe this is actually fixed in my branch, we now default to latest docker if we can't tell what openshift version is yet, which solves most of the problems in this area. (currently in master is defaults to 0.0, which is <= 3.2, and thus assumes 1.8.2 and then later has to downgrade it)
